handlers/handler-auth/src/main/java/io/servicecomb/authentication/consumer/RSACoumserTokenManager.java

```diff
@@ -0,0 +1,77 @@
+package io.servicecomb.authentication.consumer;
+
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.SignatureException;
+import java.security.spec.InvalidKeySpecException;
+import java.util.Date;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import org.apache.commons.lang3.RandomStringUtils;
+
+import io.servicecomb.authentication.RSAAuthenticationToken;
+import io.servicecomb.foundation.common.utils.RSAUtils;
+import io.servicecomb.foundation.token.AuthenticationTokenManager;
+import io.servicecomb.foundation.token.RSAKeypair4Auth;
+import io.servicecomb.serviceregistry.RegistryUtils;
+
+public class RSACoumserTokenManager implements AuthenticationTokenManager {
+
+	private ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+	
+	private RSAAuthenticationToken token;
+
+	@Override
+	public String getToken() {
+		readWriteLock.readLock().lock();
+		if(null != token && valid(token.fromat()))
+		{
+			String tokenStr = token.fromat();
+			readWriteLock.readLock().unlock();
+			return tokenStr;
+		}
+		else
+		{
+			readWriteLock.readLock().unlock();
+			return createToken();
+		}
+	}
+	
+	public String createToken()
+	{
+		String privateKey = RSAKeypair4Auth.INSTANCE.getPrivateKey();
+		readWriteLock.writeLock().lock();
+		String instanceId = RegistryUtils.getMicroserviceInstance().getInstanceId();
+		String serviceId = RegistryUtils.getMicroservice().getServiceId();
+		String randomCode = RandomStringUtils.randomAlphanumeric(128);
+		long generateTime = System.currentTimeMillis();
+		try {
+			String plain = String.format("%s@%s@%s@%s", instanceId, serviceId, generateTime, randomCode);
+			String sign = RSAUtils.sign(plain, privateKey);
+			token = RSAAuthenticationToken.fromStr(String.format("%s@%s", plain, sign));
+			return token.fromat();
+		} catch (InvalidKeyException | NoSuchAlgorithmException | InvalidKeySpecException | SignatureException e) {
+			throw new Error("create token error");
+		}
+				
+	}
+
+	/**
+	 * the TTL of Token is  24 hours
+	 * client token will expired 15 minutes early 
+	 */
+	@Override
+	public boolean valid(String token) {
+		long generateTime = RSAAuthenticationToken.fromStr(token).getGenerateTime();
+		Date expiredDate = new Date(generateTime + RSAAuthenticationToken.TOKEN_ACTIVE_TIME - 15 * 60 * 1000);
```

Comment:
why not just compare current ms and generateTime?

Reply:
because the Provider system time and Consumer system time maybe  difference ,so  that the token in Conusmer can not expired before Provider

Reply:
but your "Date now = new Date();" equals System.currentTimeMillis(), right?

Reply:
Date now = new Date() the same as Date now = new Date(System.currentTimeMillis())